include/oneapi/dpl/internal/scan_by_segment_impl.h

```diff
@@ -0,0 +1,372 @@
+                        else
+                        {
+                            __partials_acc[__group_id] = __accumulator;
+                        }
+                    }
+                });
+        });
+
+        // 2. Apply work group carry outs, calculate output indices, and load results into correct indices.
+        __exec.queue()
+            .submit([&](sycl::handler& __cgh) {
+                oneapi::dpl::__ranges::__require_access(__cgh, __keys, __out_values);
+
+                auto __partials_acc = __partials.template get_access<sycl::access_mode::read>(__cgh);
+                auto __seg_ends_acc = __seg_ends.template get_access<sycl::access_mode::read>(__cgh);
+
+                __cgh.depends_on(__wg_scan);
+
+                __dpl_sycl::__local_accessor<__val_type> __loc_partials_acc(__wgroup_size, __cgh);
+
+                __dpl_sycl::__local_accessor<__flag_type> __loc_seg_ends_acc(__wgroup_size, __cgh);
+#if _ONEDPL_COMPILE_KERNEL && _ONEDPL_KERNEL_BUNDLE_PRESENT
+                __cgh.use_kernel_bundle(__kernel2.get_kernel_bundle());
+#endif
+                __cgh.parallel_for<_SegScanKernel2>(
+#if _ONEDPL_COMPILE_KERNEL && !_ONEDPL_KERNEL_BUNDLE_PRESENT
+                    __kernel2,
+#endif
+                    sycl::nd_range<1>{__n_groups * __wgroup_size, __wgroup_size}, [=](sycl::nd_item<1> __item) {
+                        auto __group = __item.get_group();
+                        ::std::size_t __group_id = __group.get_group_id(0);
+                        ::std::size_t __global_id = __item.get_global_id();
+                        ::std::size_t __local_id = __item.get_local_id();
+                        ::std::size_t __start = __global_id * __vals_per_item;
+                        ::std::size_t __end = __dpl_sycl::__minimum<decltype(__n)>{}(__start + __vals_per_item, __n);
+
+                        ::std::size_t __wg_agg_idx = __group_id - 1;
+                        __val_type __agg_collector = __identity;
+
+                        //TODO:  just launch with one fewer group and adjust indexing since group zero can skip phase
+                        if (__group_id != 0)
+                        {
+                            // 2a. Calculate the work group's carry-in value.
+                            bool __ag_exists = false;
+                            if (__start < __n)
+                            {
+                                __ag_exists = true;
+                            }
+                            // local reductions followed by a downsweep
+                            // TODO: Generalize this value
+                            constexpr int64_t __vals_to_explore = 16;
+                            __flag_type __last_it = false;
+                            __loc_seg_ends_acc[__local_id] = false;
+                            __loc_partials_acc[__local_id] = __identity;
+
+                            for (int64_t __i = __wg_agg_idx - __vals_to_explore * __local_id;;
```

Comment:
Since we are counting downwards, it may be possible for a work item to receive a negative value of `__i`. This is checked in the nested loop below this line. We need `__i` to be a signed type for this reason which is why `int64_t` is used.